Add legacy key counter support

normalid-awa opened the pull request:
- Related:
- - #12039 
- - #25708 

Add the legacy key counter component, just like how the skinnable key counter in stable.

https://github.com/user-attachments/assets/fc115e8b-312e-4f8a-a926-2164f6fdafaf

normalid-awa committed:
Implement the overlay

normalid-awa committed:
Update the offset formula

normalid-awa committed:
Make the text are always horizontal

normalid-awa committed:
Clean up the code

normalid-awa committed:
Remove unused blending mode

normalid-awa committed:
Change the default height to match the stable

normalid-awa committed:
Replace `SkinnableSprite` with  `Sprite`

normalid-awa committed:
Fix the return button being squshed

normalid-awa committed:
* Fix the default position
* Make the font match stable style

normalid-awa committed:
Support custom input overlay color

normalid-awa committed:
Add test coverage

normalid-awa committed:
Add visual test seane

normalid-awa committed:
Match the stable animation

normalid-awa committed:
Remove logging

normalid-awa committed:
Update offset factor

normalid-awa committed:
Fix code quality

normalid-awa committed:
Fix indent problems

normalid-awa committed:
Update test assets

bdach commented:
Gotta say, this is half-baked at best.

- Default skin doesn't work correctly out-of-the-box with https://github.com/ppy/osu-resources/pull/332 assets
- Not sure what this key counter is even doing on mania, where there is not enough space to physically fit the boxes

As I know people want this, I guess I'll use this as an opportunity to port this properly on top of what's here, but I really expect a better quality of PRs going forward.

normalid-awa commented:
> Gotta say, this is half-baked at best.
> 
> - Default skin doesn't work correctly out-of-the-box with https://github.com/ppy/osu-resources/pull/332 assets
> - Not sure what this key counter is even doing on mania, where there is not enough space to physically fit the boxes
> 
> As I know people want this, I guess I'll use this as an opportunity to port this properly on top of what's here, but I really expect a better quality of PRs going forward.

https://osu.ppy.sh/wiki/en/Skinning/Interface#input-overlay
Is said that the input overlay is only available in std and catch, so it is good to just not display them in the other mode.

And i think I've missed the colour when key down, and the blending mode is not match the stable.

bdach committed:
Add better test scene

bdach committed:
Adjust layout to ballpark-match stable

I dunno what the wiki is claiming with the "24px" figure or why
but I'm not playing conversion games either. Dimensions ballparked
via screenshots captured at x768 resolution.

Also removes a weird homebrew method to keep the text upright.
There is one canonical way to do this, namely
`UprightAspectMaintainingContainer`. And the other key counters
were already using it.

bdach committed:
Simplify input overlay text colour decode (and fix incorrect default)

bdach committed:
Adjust animations further to match stable

bdach commented:
Sigh...

I am working on this branch now. Please do not push further until instructed. I'll probably force-push most of your changes anyway at this point because this code needs dire improvement anyhow.

bdach commented on osu.Game/Skinning/LegacyKeyCounter.cs:
```diff
@@ -0,0 +1,114 @@
+                {
+                    keySprite = new Sprite
+                    {
+                        Blending = BlendingParameters.Mixture,
```
dunno where this even came from, as far as i can tell nothing like this even _exists_ on stable

normalid-awa replied:
![image](https://github.com/user-attachments/assets/65c05821-9a4f-4aac-83e1-b9987c8af861)

normalid-awa replied:
The blend mode is multiplicative in stable and it did needs to be here as lot of skins rely on it

bdach replied:
"Multiplicative" doesn't require anything special. Setting `Colour` on a sprite is multiplicative already.

bdach committed:
Only add legacy key overlay to osu! and catch HUD layers

bdach commented:
I am done with this branch.

@normalid-awa please read through my changes and take note for the future. If anything is uncertain please ask.

@ppy/team-client requesting a review since I basically rewrote most of this. Stable code for your perusal at https://github.com/peppy/osu-stable-reference/blob/bb57924c1552adbed11ee3d96cdcde47cf96f2b6/osu!/Input/Drawable/InputOverlay.cs. I only matched the things that were obvious/simple to match (colours, transitions), and left well alone things that were not (I don't even know where to even begin with the stable method of specifying sizings; also stable would always add four switches and just fade out the ones that aren't used to 50% opacity; I decided I'd rather not bother with this as it's not super trivial to support this, but can if there is insistence on it).

normalid-awa commented:
> Gotta say, this is half-baked at best.
> 
> * Default skin doesn't work correctly out-of-the-box with [Add legacy skin input overlay assets osu-resources#332](https://github.com/ppy/osu-resources/pull/332) assets
> * Not sure what this key counter is even doing on mania, where there is not enough space to physically fit the boxes
> 
> As I know people want this, I guess I'll use this as an opportunity to port this properly on top of what's here, but I really expect a better quality of PRs going forward.

normalid-awa commented:
miss clicked the button,sorry

normalid-awa commented on osu.Game/Skinning/LegacySkin.cs:
```diff
@@ -442,6 +388,84 @@ protected override void ParseConfigurationStream(Stream stream)
+                }
+
+                case @"osu":
+                case @"fruits":
```
will it better to just show the `LegacyKeyCounterDisplay` at first and hide the `LegacyKeyCounterDisplay()` at mania, and taiko ruleset?

bdach replied:
I intentionally pushed this implementation over what you had. This is a bridge that will eventually have to be crossed anyway with differing implementations of legacy combo counters in osu! / the other rulesets.

normalid-awa replied:
Can we remain this https://github.com/ppy/osu/blob/0306ef4096aeb8aafd1eb49c8afefcd338a62f05/osu.Game/Skinning/LegacySkin.cs#L351-L444
and do this to hide the `LegacyKeyCounterDisplay` in mania ruleset
```diff 
:100644 100644 73c521b2ed 0000000000 M  osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaLegacySkinTransformer.cs

diff --git a/osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaLegacySkinTransformer.cs b/osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaLegacySkinTransformer.cs
index 73c521b2ed..3680306f7d 100644
--- a/osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaLegacySkinTransformer.cs
+++ b/osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaLegacySkinTransformer.cs
@@ -5,9 +5,11 @@

 using System;
 using System.Collections.Generic;
+using System.Linq;
 using osu.Framework.Audio.Sample;
 using osu.Framework.Bindables;
 using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
 using osu.Game.Audio;
 using osu.Game.Beatmaps;
 using osu.Game.Rulesets.Mania.Beatmaps;
@@ -76,6 +78,22 @@ public ManiaLegacySkinTransformer(ISkin skin, IBeatmap beatmap)

         public override Drawable GetDrawableComponent(ISkinComponentLookup lookup)
         {
+            if (lookup is SkinComponentsContainerLookup containerLookup)
+            {
+                switch (containerLookup.Target)
+                {
+                    case SkinComponentsContainerLookup.TargetArea.MainHUDComponents:
+                        var components = base.GetDrawableComponent(lookup) as Container;
+
+                        foreach (var keyCounterDisplay in components?.OfType<LegacyKeyCounterDisplay>())
+                        {
+                            keyCounterDisplay.Hide();
+                        }
+
+                        return components;
+                }
+            }
+
             switch (lookup)
             {
                 case GameplaySkinComponentLookup<HitResult> resultComponent:
```

bdach replied:
No. Fixing a unit test is an even more flimsy reasoning to keep your change. https://github.com/ppy/osu/pull/29027/commits/12a9086aa31127b4b81f5a02cbcfe190b4159b5b is how you _actually_ fix the unit test.

Unit test fixes should never be given as a rationale for an actual behavioural change in the code such as this.

bdach committed:
Fix test failure

After the legacy key counter was moved to ruleset-specific component
containers, `TestSceneSkinnableHUDOverlay` no longer had a key counter,
because it wasn't creating a ruleset-specific HUD component container
due to

    https://github.com/ppy/osu/blob/4983e5f33ed11ba3777e53face6271066ba01ab9/osu.Game/Screens/Play/HUDOverlay.cs#L131-L133

Therefore, to fix, do just enough persuading to make it create one.

Doxxz commented:
Since it's possible to have 4 keybinds in osu, could this overlay show these four keys, similar to how it was in stable?
![Captura de tela 2024-07-24 135109](https://github.com/user-attachments/assets/7877967f-0f78-47fd-89cf-819feb3f7ea7)

normalid-awa commented:
> Since it's possible to have 4 keybinds in osu, could this overlay show these four keys, similar to how it was in stable?
> ![Captura de tela 2024-07-24 135109](https://github.com/user-attachments/assets/7877967f-0f78-47fd-89cf-819feb3f7ea7)

Imo, Both bind are considered as one input (ie "z","m1" are both registered as left button"). implementing this will require a lot more changes to the base code, and would affect existing key counter.

normalid-awa commented:
> I am done with this branch.
> 
> @normalid-awa please read through my changes and take note for the future. If anything is uncertain please ask.
> 
> @ppy/team-client requesting a review since I basically rewrote most of this. Stable code for your perusal at https://github.com/peppy/osu-stable-reference/blob/bb57924c1552adbed11ee3d96cdcde47cf96f2b6/osu!/Input/Drawable/InputOverlay.cs.
> 
> I only matched the things that were obvious/simple to match (colours, transitions), and left well alone things that were not (I don't even know where to even begin with the stable method of specifying sizings; also stable would always add four switches and just fade out the ones that aren't used to 50% opacity; I decided I'd rather not bother with this as it's not super trivial to support this, but can if there is insistence on it).
> 
> Oh, and [c3dae81](https://github.com/ppy/osu/commit/c3dae81935a18eb62ebbfa7fdc2c36af5fd9fe2f) probably deserves particular scrutiny because it is unprecedented.

@bdach  May I ask about the progress of the review?

bdach commented:
I'm not the one to ask. I asked others for review. I'm not reviewing this myself because I changed too much.

peppy committed:
Merge branch 'master' into feature/skin/legacy-input-overlay

peppy commented on osu.Game/Skinning/LegacySkin.cs:
```diff
@@ -437,6 +388,84 @@ protected override void ParseConfigurationStream(Stream stream)
 
+        private static DefaultSkinComponentsContainer? createDefaultHUDComponents(SkinComponentsContainerLookup containerLookup)
+        {
+            switch (containerLookup.Ruleset?.ShortName)
```
So.. this hinges on the fact that there's guaranteed to be a `SkinComponentsContainerLookup` request for `null` (global) then per ruleset?

I'm sure you did already consider this, but would it be better or worse to put the ruleset version in `OsuLegacySkinTransformer` etc.?

normalid-awa replied:
Seems to be discussed before. [ref](https://github.com/ppy/osu/pull/29027#discussion_r1689938442)
But I'm still have the preferred as put the legacy key per-ruleset.

peppy replied:
```diff
diff --git a/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs b/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs
index d1ef47cf17..550ce0b53c 100644
--- a/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs
+++ b/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs
@@ -6,6 +6,7 @@
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Containers;
 using osu.Game.Skinning;
+using osuTK;
 using osuTK.Graphics;
 
 namespace osu.Game.Rulesets.Catch.Skinning.Legacy
@@ -28,11 +29,15 @@ public CatchLegacySkinTransformer(ISkin skin)
 
         public override Drawable? GetDrawableComponent(ISkinComponentLookup lookup)
         {
-            if (lookup is SkinComponentsContainerLookup containerLookup)
+            switch (lookup)
             {
-                switch (containerLookup.Target)
-                {
-                    case SkinComponentsContainerLookup.TargetArea.MainHUDComponents:
+                case SkinComponentsContainerLookup containerLookup:
+                    if (containerLookup.Target != SkinComponentsContainerLookup.TargetArea.MainHUDComponents)
+                        return base.GetDrawableComponent(lookup);
+
+                    // Modifications for global components.
+                    if (containerLookup.Ruleset == null)
+                    {
                         var components = base.GetDrawableComponent(lookup) as Container;
 
                         if (providesComboCounter && components != null)
@@ -44,60 +49,80 @@ public CatchLegacySkinTransformer(ISkin skin)
                         }
 
                         return components;
-                }
-            }
+                    }
 
-            if (lookup is CatchSkinComponentLookup catchSkinComponent)
-            {
-                switch (catchSkinComponent.Component)
-                {
-                    case CatchSkinComponents.Fruit:
-                        if (hasPear)
-                            return new LegacyFruitPiece();
+                    // Our own ruleset components.
+                    return new DefaultSkinComponentsContainer(container =>
+                    {
+                        var keyCounter = container.OfType<LegacyKeyCounterDisplay>().FirstOrDefault();
 
-                        return null;
+                        if (keyCounter != null)
+                        {
+                            // set the anchor to top right so that it won't squash to the return button to the top
+                            keyCounter.Anchor = Anchor.CentreRight;
+                            keyCounter.Origin = Anchor.CentreRight;
+                            keyCounter.X = 0;
+                            // 340px is the default height inherit from stable
+                            keyCounter.Y = container.ToLocalSpace(new Vector2(0, container.ScreenSpaceDrawQuad.Centre.Y - 340f)).Y;
+                        }
+                    })
+                    {
+                        Children = new Drawable[]
+                        {
+                            new LegacyKeyCounterDisplay(),
+                        }
+                    };
 
-                    case CatchSkinComponents.Banana:
-                        if (GetTexture("fruit-bananas") != null)
-                            return new LegacyBananaPiece();
+                case CatchSkinComponentLookup catchSkinComponent:
+                    switch (catchSkinComponent.Component)
+                    {
+                        case CatchSkinComponents.Fruit:
+                            if (hasPear)
+                                return new LegacyFruitPiece();
 
-                        return null;
+                            return null;
 
-                    case CatchSkinComponents.Droplet:
-                        if (GetTexture("fruit-drop") != null)
-                            return new LegacyDropletPiece();
+                        case CatchSkinComponents.Banana:
+                            if (GetTexture("fruit-bananas") != null)
+                                return new LegacyBananaPiece();
 
-                        return null;
+                            return null;
 
-                    case CatchSkinComponents.Catcher:
-                        decimal version = GetConfig<SkinConfiguration.LegacySetting, decimal>(SkinConfiguration.LegacySetting.Version)?.Value ?? 1;
+                        case CatchSkinComponents.Droplet:
+                            if (GetTexture("fruit-drop") != null)
+                                return new LegacyDropletPiece();
 
-                        if (version < 2.3m)
-                        {
-                            if (hasOldStyleCatcherSprite())
-                                return new LegacyCatcherOld();
-                        }
+                            return null;
 
-                        if (hasNewStyleCatcherSprite())
-                            return new LegacyCatcherNew();
+                        case CatchSkinComponents.Catcher:
+                            decimal version = GetConfig<SkinConfiguration.LegacySetting, decimal>(SkinConfiguration.LegacySetting.Version)?.Value ?? 1;
 
-                        return null;
+                            if (version < 2.3m)
+                            {
+                                if (hasOldStyleCatcherSprite())
+                                    return new LegacyCatcherOld();
+                            }
 
-                    case CatchSkinComponents.CatchComboCounter:
-                        if (providesComboCounter)
-                            return new LegacyCatchComboCounter();
+                            if (hasNewStyleCatcherSprite())
+                                return new LegacyCatcherNew();
 
-                        return null;
+                            return null;
 
-                    case CatchSkinComponents.HitExplosion:
-                        if (hasOldStyleCatcherSprite() || hasNewStyleCatcherSprite())
-                            return new LegacyHitExplosion();
+                        case CatchSkinComponents.CatchComboCounter:
+                            if (providesComboCounter)
+                                return new LegacyCatchComboCounter();
 
-                        return null;
+                            return null;
+
+                        case CatchSkinComponents.HitExplosion:
+                            if (hasOldStyleCatcherSprite() || hasNewStyleCatcherSprite())
+                                return new LegacyHitExplosion();
+
+                            return null;
 
-                    default:
-                        throw new UnsupportedSkinComponentException(lookup);
-                }
+                        default:
+                            throw new UnsupportedSkinComponentException(lookup);
+                    }
             }
 
             return base.GetDrawableComponent(lookup);
diff --git a/osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs b/osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs
index d2ebc68c52..0923ebc832 100644
--- a/osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs
+++ b/osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs
@@ -2,6 +2,7 @@
 // See the LICENCE file in the repository root for full licence text.
 
 using System;
+using System.Linq;
 using osu.Framework.Bindables;
 using osu.Framework.Graphics;
 using osu.Game.Rulesets.Osu.Objects;
@@ -41,139 +42,173 @@ public OsuLegacySkinTransformer(ISkin skin)
 
         public override Drawable? GetDrawableComponent(ISkinComponentLookup lookup)
         {
-            if (lookup is OsuSkinComponentLookup osuComponent)
+            switch (lookup)
             {
-                switch (osuComponent.Component)
-                {
-                    case OsuSkinComponents.FollowPoint:
-                        return this.GetAnimation("followpoint", true, true, true, startAtCurrentTime: false, maxSize: new Vector2(OsuHitObject.OBJECT_RADIUS * 2, OsuHitObject.OBJECT_RADIUS));
+                case SkinComponentsContainerLookup containerLookup:
+                    // Only handle per ruleset defaults here.
+                    if (containerLookup.Ruleset == null)
+                        return base.GetDrawableComponent(lookup);
 
-                    case OsuSkinComponents.SliderScorePoint:
-                        return this.GetAnimation("sliderscorepoint", false, false, maxSize: OsuHitObject.OBJECT_DIMENSIONS);
+                    switch (containerLookup.Target)
+                    {
+                        case SkinComponentsContainerLookup.TargetArea.MainHUDComponents:
+                            return new DefaultSkinComponentsContainer(container =>
+                            {
+                                var keyCounter = container.OfType<LegacyKeyCounterDisplay>().FirstOrDefault();
+
+                                if (keyCounter != null)
+                                {
+                                    // set the anchor to top right so that it won't squash to the return button to the top
+                                    keyCounter.Anchor = Anchor.CentreRight;
+                                    keyCounter.Origin = Anchor.CentreRight;
+                                    keyCounter.X = 0;
+                                    // 340px is the default height inherit from stable
+                                    keyCounter.Y = container.ToLocalSpace(new Vector2(0, container.ScreenSpaceDrawQuad.Centre.Y - 340f)).Y;
+                                }
+                            })
+                            {
+                                Children = new Drawable[]
+                                {
+                                    new LegacyKeyCounterDisplay(),
+                                }
+                            };
+                    }
 
-                    case OsuSkinComponents.SliderFollowCircle:
-                        var followCircleContent = this.GetAnimation("sliderfollowcircle", true, true, true, maxSize: MAX_FOLLOW_CIRCLE_AREA_SIZE);
-                        if (followCircleContent != null)
-                            return new LegacyFollowCircle(followCircleContent);
+                    return null;
 
-                        return null;
+                case OsuSkinComponentLookup osuComponent:
+                    switch (osuComponent.Component)
+                    {
+                        case OsuSkinComponents.FollowPoint:
+                            return this.GetAnimation("followpoint", true, true, true, startAtCurrentTime: false, maxSize: new Vector2(OsuHitObject.OBJECT_RADIUS * 2, OsuHitObject.OBJECT_RADIUS));
 
-                    case OsuSkinComponents.SliderBall:
-                        if (GetTexture("sliderb") != null || GetTexture("sliderb0") != null)
-                            return new LegacySliderBall(this);
+                        case OsuSkinComponents.SliderScorePoint:
+                            return this.GetAnimation("sliderscorepoint", false, false, maxSize: OsuHitObject.OBJECT_DIMENSIONS);
 
-                        return null;
+                        case OsuSkinComponents.SliderFollowCircle:
+                            var followCircleContent = this.GetAnimation("sliderfollowcircle", true, true, true, maxSize: MAX_FOLLOW_CIRCLE_AREA_SIZE);
+                            if (followCircleContent != null)
+                                return new LegacyFollowCircle(followCircleContent);
 
-                    case OsuSkinComponents.SliderBody:
-                        if (hasHitCircle.Value)
-                            return new LegacySliderBody();
+                            return null;
 
-                        return null;
+                        case OsuSkinComponents.SliderBall:
+                            if (GetTexture("sliderb") != null || GetTexture("sliderb0") != null)
+                                return new LegacySliderBall(this);
 
-                    case OsuSkinComponents.SliderTailHitCircle:
-                        if (hasHitCircle.Value)
-                            return new LegacyMainCirclePiece("sliderendcircle", false);
+                            return null;
 
-                        return null;
+                        case OsuSkinComponents.SliderBody:
+                            if (hasHitCircle.Value)
+                                return new LegacySliderBody();
 
-                    case OsuSkinComponents.SliderHeadHitCircle:
-                        if (hasHitCircle.Value)
-                            return new LegacySliderHeadHitCircle();
+                            return null;
 
-                        return null;
+                        case OsuSkinComponents.SliderTailHitCircle:
+                            if (hasHitCircle.Value)
+                                return new LegacyMainCirclePiece("sliderendcircle", false);
 
-                    case OsuSkinComponents.ReverseArrow:
-                        if (hasHitCircle.Value)
-                            return new LegacyReverseArrow();
+                            return null;
 
-                        return null;
+                        case OsuSkinComponents.SliderHeadHitCircle:
+                            if (hasHitCircle.Value)
+                                return new LegacySliderHeadHitCircle();
 
-                    case OsuSkinComponents.HitCircle:
-                        if (hasHitCircle.Value)
-                            return new LegacyMainCirclePiece();
+                            return null;
 
-                        return null;
+                        case OsuSkinComponents.ReverseArrow:
+                            if (hasHitCircle.Value)
+                                return new LegacyReverseArrow();
 
-                    case OsuSkinComponents.Cursor:
-                        if (GetTexture("cursor") != null)
-                            return new LegacyCursor(this);
+                            return null;
 
-                        return null;
+                        case OsuSkinComponents.HitCircle:
+                            if (hasHitCircle.Value)
+                                return new LegacyMainCirclePiece();
 
-                    case OsuSkinComponents.CursorTrail:
-                        if (GetTexture("cursortrail") != null)
-                            return new LegacyCursorTrail(this);
+                            return null;
 
-                        return null;
+                        case OsuSkinComponents.Cursor:
+                            if (GetTexture("cursor") != null)
+                                return new LegacyCursor(this);
 
-                    case OsuSkinComponents.CursorRipple:
-                        if (GetTexture("cursor-ripple") != null)
-                        {
-                            var ripple = this.GetAnimation("cursor-ripple", false, false);
+                            return null;
 
-                            // In stable this element was scaled down to 50% and opacity 20%, but this makes the elements WAY too big and inflexible.
-                            // If anyone complains about these not being applied, this can be uncommented.
-                            //
-                            // But if no one complains I'd rather fix this in lazer. Wiki documentation doesn't mention size,
-                            // so we might be okay.
-                            //
-                            // if (ripple != null)
-                            // {
-                            //     ripple.Scale = new Vector2(0.5f);
-                            //     ripple.Alpha = 0.2f;
-                            // }
+                        case OsuSkinComponents.CursorTrail:
+                            if (GetTexture("cursortrail") != null)
+                                return new LegacyCursorTrail(this);
 
-                            return ripple;
-                        }
+                            return null;
 
-                        return null;
+                        case OsuSkinComponents.CursorRipple:
+                            if (GetTexture("cursor-ripple") != null)
+                            {
+                                var ripple = this.GetAnimation("cursor-ripple", false, false);
+
+                                // In stable this element was scaled down to 50% and opacity 20%, but this makes the elements WAY too big and inflexible.
+                                // If anyone complains about these not being applied, this can be uncommented.
+                                //
+                                // But if no one complains I'd rather fix this in lazer. Wiki documentation doesn't mention size,
+                                // so we might be okay.
+                                //
+                                // if (ripple != null)
+                                // {
+                                //     ripple.Scale = new Vector2(0.5f);
+                                //     ripple.Alpha = 0.2f;
+                                // }
+
+                                return ripple;
+                            }
 
-                    case OsuSkinComponents.CursorParticles:
-                        if (GetTexture("star2") != null)
-                            return new LegacyCursorParticles();
+                            return null;
 
-                        return null;
+                        case OsuSkinComponents.CursorParticles:
+                            if (GetTexture("star2") != null)
+                                return new LegacyCursorParticles();
 
-                    case OsuSkinComponents.CursorSmoke:
-                        if (GetTexture("cursor-smoke") != null)
-                            return new LegacySmokeSegment();
+                            return null;
 
-                        return null;
+                        case OsuSkinComponents.CursorSmoke:
+                            if (GetTexture("cursor-smoke") != null)
+                                return new LegacySmokeSegment();
 
-                    case OsuSkinComponents.HitCircleText:
-                        if (!this.HasFont(LegacyFont.HitCircle))
                             return null;
 
-                        const float hitcircle_text_scale = 0.8f;
-                        return new LegacySpriteText(LegacyFont.HitCircle)
-                        {
-                            // stable applies a blanket 0.8x scale to hitcircle fonts
-                            Scale = new Vector2(hitcircle_text_scale),
-                            MaxSizePerGlyph = OsuHitObject.OBJECT_DIMENSIONS * 2 / hitcircle_text_scale,
-                        };
+                        case OsuSkinComponents.HitCircleText:
+                            if (!this.HasFont(LegacyFont.HitCircle))
+                                return null;
 
-                    case OsuSkinComponents.SpinnerBody:
-                        bool hasBackground = GetTexture("spinner-background") != null;
+                            const float hitcircle_text_scale = 0.8f;
+                            return new LegacySpriteText(LegacyFont.HitCircle)
+                            {
+                                // stable applies a blanket 0.8x scale to hitcircle fonts
+                                Scale = new Vector2(hitcircle_text_scale),
+                                MaxSizePerGlyph = OsuHitObject.OBJECT_DIMENSIONS * 2 / hitcircle_text_scale,
+                            };
 
-                        if (GetTexture("spinner-top") != null && !hasBackground)
-                            return new LegacyNewStyleSpinner();
-                        else if (hasBackground)
-                            return new LegacyOldStyleSpinner();
+                        case OsuSkinComponents.SpinnerBody:
+                            bool hasBackground = GetTexture("spinner-background") != null;
 
-                        return null;
+                            if (GetTexture("spinner-top") != null && !hasBackground)
+                                return new LegacyNewStyleSpinner();
+                            else if (hasBackground)
+                                return new LegacyOldStyleSpinner();
 
-                    case OsuSkinComponents.ApproachCircle:
-                        if (GetTexture(@"approachcircle") != null)
-                            return new LegacyApproachCircle();
+                            return null;
 
-                        return null;
+                        case OsuSkinComponents.ApproachCircle:
+                            if (GetTexture(@"approachcircle") != null)
+                                return new LegacyApproachCircle();
 
-                    default:
-                        throw new UnsupportedSkinComponentException(lookup);
-                }
-            }
+                            return null;
 
-            return base.GetDrawableComponent(lookup);
+                        default:
+                            throw new UnsupportedSkinComponentException(lookup);
+                    }
+
+                default:
+                    return base.GetDrawableComponent(lookup);
+            }
         }
 
         public override IBindable<TValue>? GetConfig<TLookup, TValue>(TLookup lookup)
diff --git a/osu.Game/Skinning/LegacySkin.cs b/osu.Game/Skinning/LegacySkin.cs
index 5c3af4d52e..b1b171eef9 100644
--- a/osu.Game/Skinning/LegacySkin.cs
+++ b/osu.Game/Skinning/LegacySkin.cs
@@ -23,7 +23,6 @@
 using osu.Game.Rulesets.Scoring;
 using osu.Game.Screens.Play.HUD;
 using osu.Game.Screens.Play.HUD.HitErrorMeters;
-using osuTK;
 using osuTK.Graphics;
 
 namespace osu.Game.Skinning
@@ -429,84 +428,6 @@ protected override void ParseConfigurationStream(Stream stream)
             return null;
         }
 
-        private static DefaultSkinComponentsContainer? createDefaultHUDComponents(SkinComponentsContainerLookup containerLookup)
-        {
-            switch (containerLookup.Ruleset?.ShortName)
-            {
-                case null:
-                {
-                    return new DefaultSkinComponentsContainer(container =>
-                    {
-                        var score = container.OfType<LegacyScoreCounter>().FirstOrDefault();
-                        var accuracy = container.OfType<GameplayAccuracyCounter>().FirstOrDefault();
-
-                        if (score != null && accuracy != null)
-                        {
-                            accuracy.Y = container.ToLocalSpace(score.ScreenSpaceDrawQuad.BottomRight).Y;
-                        }
-
-                        var songProgress = container.OfType<LegacySongProgress>().FirstOrDefault();
-
-                        if (songProgress != null && accuracy != null)
-                        {
-                            songProgress.Anchor = Anchor.TopRight;
-                            songProgress.Origin = Anchor.CentreRight;
-                            songProgress.X = -accuracy.ScreenSpaceDeltaToParentSpace(accuracy.ScreenSpaceDrawQuad.Size).X - 18;
-                            songProgress.Y = container.ToLocalSpace(accuracy.ScreenSpaceDrawQuad.TopLeft).Y + (accuracy.ScreenSpaceDeltaToParentSpace(accuracy.ScreenSpaceDrawQuad.Size).Y / 2);
-                        }
-
-                        var hitError = container.OfType<HitErrorMeter>().FirstOrDefault();
-
-                        if (hitError != null)
-                        {
-                            hitError.Anchor = Anchor.BottomCentre;
-                            hitError.Origin = Anchor.CentreLeft;
-                            hitError.Rotation = -90;
-                        }
-                    })
-                    {
-                        Children = new Drawable[]
-                        {
-                            new LegacyComboCounter(),
-                            new LegacyScoreCounter(),
-                            new LegacyAccuracyCounter(),
-                            new LegacySongProgress(),
-                            new LegacyHealthDisplay(),
-                            new BarHitErrorMeter(),
-                        }
-                    };
-                }
-
-                case @"osu":
-                case @"fruits":
-                {
-                    return new DefaultSkinComponentsContainer(container =>
-                    {
-                        var keyCounter = container.OfType<LegacyKeyCounterDisplay>().FirstOrDefault();
-
-                        if (keyCounter != null)
-                        {
-                            // set the anchor to top right so that it won't squash to the return button to the top
-                            keyCounter.Anchor = Anchor.CentreRight;
-                            keyCounter.Origin = Anchor.CentreRight;
-                            keyCounter.X = 0;
-                            // 340px is the default height inherit from stable
-                            keyCounter.Y = container.ToLocalSpace(new Vector2(0, container.ScreenSpaceDrawQuad.Centre.Y - 340f)).Y;
-                        }
-                    })
-                    {
-                        Children = new Drawable[]
-                        {
-                            new LegacyKeyCounterDisplay(),
-                        }
-                    };
-                }
-
-                default:
-                    return null;
-            }
-        }
-
         private Texture? getParticleTexture(HitResult result)
         {
             switch (result)

```

This seems to work well, and follow more closely to the expectations I'd have from the skinning structure (we already have these transformer classes for rulesets modification like this).

peppy replied:
I had to update the diff, accidentally shared a partial diff, sorry.

peppy replied:
I just committed it, the diff is huge and whatever.

normalid-awa replied:
> I just committed it, the diff is huge and whatever.
> 
> [725dc4d](https://github.com/ppy/osu/commit/725dc4de9b80773fe057059eacab9ae244ed21ac) Highly recommend hiding whitespace changes.

Looks good, Yours approach looks similar to the original thought when I was trying to bring the legacy key counter to the default. But my approach is to instantiate the key counter globally, and disable them in taiko/mania, or whatever the ruleset doesn't compact with the legacy key counter. But seems like yours approach looks more clean and readable!

bdach replied:
I don't buy that commit.

Why are _legacy skin defaults_ being handled by _skin-agnostic ruleset transformers_?

I'm not even gonna comment on the code itself because it's basically illegible to me and breaks my entire mental model of how any of this is supposed to work. Ruleset transformers doing something for null ruleset, then some fallback to base, and then some manual operating on a skin that may or may not contain a legacy counter. How is that better than what I had?

peppy commented on osu.Game/Skinning/LegacyKeyCounterDisplay.cs:
```diff
@@ -0,0 +1,93 @@
+
+            AddRangeInternal(new Drawable[]
+            {
+                backgroundSprite = new Sprite
```
Because the background isn't contained within `KeyFlow`, the background will show regardless of the "always show key overlay" setting:

![osu! 2024-08-06 at 05 42 11](https://github.com/user-attachments/assets/1fe0ecdd-ba9b-42a9-9e24-8f44342f4b5f)

This is due to 

https://github.com/ppy/osu/blob/798e677c092a4ad02e0fe0a81163c14b57ff90d4/osu.Game/Screens/Play/HUD/KeyCounterDisplay.cs#L37-L47

Should be simple enough to fix, I may look at it subsequently

peppy replied:
Fixed in https://github.com/peppy/osu/commit/8619bbb9435c82018ffc50ab15a61430ae77146a.

frenzibyte commented:
I'm also noticing the key overlay to display incorrectly when "beatmap skins" is enabled:

https://github.com/user-attachments/assets/6e60359d-a412-45a7-b0bd-9ebf852324e1

Skin: [-         _CK_ WhiteCat 2.0 _ new (cyperdark).osk.zip](https://github.com/user-attachments/files/16505926/-._CK_.WhiteCat.2.0._.new.cyperdark.osk.zip)

Happens on any beatmap as far as I'm aware.

peppy commented:
> I'm also noticing the key overlay to display incorrectly when "beatmap skins" is enabled

Looks like this config lookup needs to use `getCustomColour` instead of `legacySettingLookup`. I'll take a quick look at fixing this too.

peppy committed:
Sanitise font sizes / weights

peppy committed:
Fix incorrect colour fallback handling

Adds a note about `GetConfig` being stupid.

peppy committed:
Fix legacy key counter's background being visible when intended to be hidden

peppy commented:
@normalid-awa please refrain from committing.

peppy committed:
Use transformers for per-skin key counter implementation

peppy reviewed (approved):
Seems probably okay now

smoogipoo commented on osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs:
```diff
@@ -41,139 +42,178 @@ public OsuLegacySkinTransformer(ISkin skin)
+                case SkinComponentsContainerLookup containerLookup:
+                    // Only handle per ruleset defaults here.
+                    if (containerLookup.Ruleset == null)
+                        return base.GetDrawableComponent(lookup);
 
-                    case OsuSkinComponents.SliderScorePoint:
-                        return this.GetAnimation("sliderscorepoint", false, false, maxSize: OsuHitObject.OBJECT_DIMENSIONS);
+                    // Skin has configuration.
+                    if (base.GetDrawableComponent(lookup) is Drawable d)
+                        return d;
```
I will point out that this was brought up as one of the primary issues with my prior [ruleset component skinning PR](https://github.com/ppy/osu/pull/27773#discussion_r1549287482) that resulted in me basically rewriting how that PR is done.

TheFantasticLoki commented:
IK this issue is closed but I don't want to make another issue or discussion and get yelled at for not just using an existing issue.

I believe there is some kind of problem with the implementation of this as I added the legacy key counter to my skin and it was working fine for a few days and then all of a sudden it just, stopped working. It still shows up if your using autoplay which skin editor will usually just turn on while editing the skin, but during actual gameplay it doesn't show up and if you open the editor it just shows as a blank square.

Working with auto:
<img width="1919" height="1080" alt="image" src="https://github.com/user-attachments/assets/ed44966e-1e7b-4ae8-a42a-61a7f92b9efa" />

Not working on regular gameplay:
<img width="1920" height="1083" alt="image" src="https://github.com/user-attachments/assets/f97f8bed-e113-46fb-baaf-76bdc92ca64d" />

Just shows as a tiny little square in editor without auto turned on:
<img width="1918" height="1080" alt="image" src="https://github.com/user-attachments/assets/a9ad92a8-e311-4b05-8257-013b34a8e7e0" />

TheFantasticLoki commented:
> @TheFantasticLoki* <img alt="image" width="435" height="140" src="https://private-user-images.githubusercontent.com/24654479/474621203-0241f3e5-07fa-425e-b0ae-bd99480b8e31.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ0MTEwMzIsIm5iZiI6MTc1NDQxMDczMiwicGF0aCI6Ii8yNDY1NDQ3OS80NzQ2MjEyMDMtMDI0MWYzZTUtMDdmYS00MjVlLWIwYWUtYmQ5OTQ4MGI4ZTMxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODA1VDE2MTg1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMyZGI3ZmVmNWJlNGIxZTkyZjllMmRjN2I0MDI4M2VjOGU4MDQ1MTY4MGU5MzM5MDEwNTBiZWMzMzg4MjMyNWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Wvus_vpNwu2S5B9oDAJUg6vtkY1XRnQTANPbnqOH7dQ">

Thank you.... I'm not even gonna question how this option was turned on for like a week and then randomly got disabled causing this missunderstanding.
